designs/0005-static-matrices.md

```diff
@@ -0,0 +1,252 @@
+
+Now that `varmat` variables can be assigned, it is possible that the compiler silently replace every
+`matvar` in Stan with a `varmat`, and none of the standard Stan use cases would be affected.
+
+There are four major implementation strategies available at the Stanc3 level.
+
+1. Implement a new Stan type to differentiate explicitly between `matvar` and `varmat` types
+
+2. Use `matvar` types for any function in the Stan program that does not support `varmat`
+
+3. Use `varmat` types by default, but if variable assignment is detected, or a `varmat` is used
+in an unsupported function convert to and from `matvar` automatically.
+
+4. Use `varmat` types everywhere
+
+The original proposal was leaning towards #1 because it most naturally corresponds to the Math
+implementation. However, this leaves the problem of explicitly declaring `varmat` to the user. To
+avoid duplicating the user's manual and function reference for `varmat` and `matvar` functions, there
+would also need to be some automatic, under the hood conversions between the types. This automatic
+converting would make it so that if a `varmat` is forgotten anywhere, the entire program will be
+slowed down. This sort of explicit typing shouldn't be too much of a problem in development models
+or generated Stan code, but in large, user-written models this would be error prone to implement
+and maintain.
+
+The advantage of #2 is that it would allow programs that can be written using only `varmat` functions
+to be very fast (and not even worry about any conversion to `matvar` or not). The problem with this
+implementation is that, the difference between `varmat` and `matvar` would again need documented
+at the user level and some notation would need added to the user manual to indicate if a function
+is or is not available for optimization.
+
+The advantage of #3, similarly to #2, is that `varmat` and `matvar` types could be handled
+automatically by the compiler. If the user requires functions only available for `matvar`,
+there will be some slowdown as conversions between `varmat` and `matvar` are handled. However,
+if the conversions are only used in relatively cheap parts of the computation, then the expensive
+parts of the calculation can still try to take advantage of `varmat` types. The difficulty here
+is compiler implementation which would need to be more advanced to detect when it could use
+`varmat` and when it would need to convert to `matvar`.
+
+The advantage of #4 is simplicitly of use and simplicitly of compiler implementation. If `varmat`
+types are used everywhere, and there is no attempt to balance against the advantages of
+`varmat` against `matvar`, then the main difficulty will just be making sure to convert `varmat`
+variables to `matvar` variables and back again when used in functions that do not support `varmat`.
+In the limit that all functions can be written to support `varmat`, then #2, #3, and #4 become
+equivalent.
```

Comment:
4 was kind of meant more in the sense switch to `varmat` when we get everything in Math and scratch `matvar`.

Reply:
This is simpler than #3 though right? This is like #3, except we don't try to be smart about making stuff a `varmat` or a `matvar`. We just, as much as possible, make things `varmat` types.

If this doesn't make sense to you as a proposal we should change it though lol, cuz this was my interpretation of a list you made :P.

Reply:
No, I meant it as "wait till everything it varmat" and then just switch to it. It's not a good option, I just wanted to present all cases.

Reply:
Okay, I'll remove this one then. If it's not realistic and we don't want to do it, we shouldn't worry about listing it.

We haven't truly listed all options, so let's not worry about that (for instance, we could also stick with `_varmat`).

Reply:
Ok yeah, good point.